### ISSUE TITLE
Custom memory management for genes

### ISSUE THREAD

**Issue by matiasmicheletto** [Arash-codedev/openGA#19]
Hi, thanks for this great tool and contribution.

This issue may be more related to C++ than to the library itself, but I hope that someone can give me a hand here.

I'm using this library to compute an optimal binary array with many elements (up to 100,000, or more depending on computer resources). 

To give a few more details, these arrays models some networks of nodes, and each element indicates if the network has a gateway (GW) in that node or not. The goal is to minimize the number of GW, that is, the number of elements that are "true". So the solution length is user-defined, and stored in a variable (```_net_size```).

This is the individual (or solution) model:

```cpp
struct MySolution {
    bool* isGW; // Element: true -> node has GW, false -> node has no GW

    string to_string() const { // Only print number of GW
    unsigned int gwCnt = 0; // GW counter
    for(unsigned int i = 0; i < _net_size; i++){
        if(isGW[i]){
            gwCnt++;
        }
    }
    return std::to_string(gwCnt)+" GWs";
    }
};
```

To initialize the genes of a certain individual, I'm using this method:

```cpp
void init_genes(
    MySolution& p, 
    const function<double(void)> &rnd01) 
{
	p.isGW = (bool*) malloc(sizeof(bool)*_net_size); // (can I free this later??)
	for(unsigned int i = 0; i < _net_size; i++)
		p.isGW[i] = rnd01() < 0.001; // 1 gw every 1000 nodes
}
```

Now, the problem comes when implementing the mutation and crossover operators, because I think I am accessing and modifying ```X_base``` (in ```mutate```), or ```X1``` and ```X2``` parents values (in ```crossover```).

However, if I initialize a new individual every time, this is, allocating memory, on ```mutate``` or ```crossover``` functions, then I believe the memory usage will grow up indefinitely (memory leakage).

```cpp
MySolution mutate(
    const MySolution& X_base, 
    const function<double(void)> &rnd01, 
    double shrink_scale) 
{
    MySolution X_new = X_base; // Should I use malloc here and copy values of X_base?

    for(unsigned int i = 0; i < _net_size; i++){
        if(rnd01() < _mut_prob){ // _mut_prob equals 1/_net_size
            X_new.isGW[i] = !X_new.isGW[i]; // Switch gene value
        }
    }

	return X_new;
}

MySolution crossover(
    const MySolution& X1, 
    const MySolution& X2,
    const function<double(void)> &rnd01) 
{
    MySolution X_new = X1; // Should I use malloc here and copy values of X1?

    // Random crossover point
    unsigned int x_point = (unsigned int) (rnd01()*(double)_net_size);
    for(unsigned int i = x_point; i < _net_size; i++){
        X_new.isGW[i] = X2.isGW[i];
    }

	return X_new;
}
```

So, when running this code, I get a ```best``` value that is correct, but the ```best_genes``` are not, because you can clearly see that GWs count is getting bigger every generation:

```cpp
void SO_report_generation(int generation_number, const EA::GenerationType<Chromosome,MiddleCost> &last_generation, const Chromosome& best_genes) {
    cout<<"Generation ["<<generation_number<<"], "
	<<"Best = "<<last_generation.best_total_cost<<", "
	<<"Average = "<<last_generation.average_cost<<", "
	<<"Best genes = "<<best_genes.to_string()<<", "
        <<"Best index = "<<last_generation.best_chromosome_index<<", "
	<<"Exe_time = "<<last_generation.exe_time
	<<endl;
}
```

```
Generation [0], Best = 5, Average = 164290, Best genes = 5 GWs, Best index = 29, Exe_time = 0.23819
Generation [1], Best = 5, Average = 101229, Best genes = 7 GWs, Best index = 0, Exe_time = 0.0577664
Generation [2], Best = 5, Average = 93488.9, Best genes = 5 GWs, Best index = 4, Exe_time = 0.0537195
Generation [3], Best = 5, Average = 8.62, Best genes = 4 GWs, Best index = 11, Exe_time = 0.043536
Generation [4], Best = 5, Average = 8.6, Best genes = 7 GWs, Best index = 7, Exe_time = 0.0385683
Generation [5], Best = 5, Average = 7.62, Best genes = 12 GWs, Best index = 48, Exe_time = 0.072586
Generation [6], Best = 5, Average = 7.34, Best genes = 12 GWs, Best index = 4, Exe_time = 0.0605616
Generation [7], Best = 5, Average = 7.34, Best genes = 9 GWs, Best index = 4, Exe_time = 0.0445714
Generation [8], Best = 5, Average = 6.8, Best genes = 10 GWs, Best index = 4, Exe_time = 0.066782
Generation [9], Best = 5, Average = 6.8, Best genes = 12 GWs, Best index = 4, Exe_time = 0.0528507
Generation [10], Best = 5, Average = 6.44, Best genes = 15 GWs, Best index = 4, Exe_time = 0.0641599
Generation [11], Best = 5, Average = 6.62, Best genes = 13 GWs, Best index = 4, Exe_time = 0.056146
Generation [12], Best = 5, Average = 6.44, Best genes = 16 GWs, Best index = 4, Exe_time = 0.0741491
Generation [13], Best = 5, Average = 6.8, Best genes = 15 GWs, Best index = 4, Exe_time = 0.0670279
Generation [14], Best = 5, Average = 6.62, Best genes = 18 GWs, Best index = 4, Exe_time = 0.0729863
Generation [15], Best = 5, Average = 6.98, Best genes = 22 GWs, Best index = 4, Exe_time = 0.0699864
Generation [16], Best = 5, Average = 6.62, Best genes = 13 GWs, Best index = 4, Exe_time = 0.0863633
Generation [17], Best = 5, Average = 6.08, Best genes = 20 GWs, Best index = 4, Exe_time = 0.0677662
Generation [18], Best = 5, Average = 6.08, Best genes = 20 GWs, Best index = 4, Exe_time = 0.0694805
Generation [19], Best = 5, Average = 6.08, Best genes = 21 GWs, Best index = 4, Exe_time = 0.0636514
Generation [20], Best = 5, Average = 5.9, Best genes = 24 GWs, Best index = 4, Exe_time = 0.0818476
```

That is the reason I believe I'm screwing up parents on ```crossover``` method.

So, in short, is it possible to perform the corresponding ```free()``` for the case of using ```malloc()```? How should I fix this isssue?

Thanks in advance!

**Comment by matiasmicheletto** [comment]
Just tried using memory allocation on every ```mutate``` and ```crossover``` call, and memory usage does not look to increase too much. This is the final code for these two methods:

```cpp
void copy_genes(const MySolution& from, MySolution& to) {
    for(unsigned int i = 0; i < _enddevices; i++)
        to.isGW[i] = from.isGW[i];
}

MySolution mutate(const MySolution& X_base, const function<double(void)> &rnd01, double shrink_scale) {
	MySolution X_new;
    init_genes(X_new, rnd01);
    copy_genes(X_base, X_new);

    for(unsigned int i = 0; i < _enddevices; i++){
        if(rnd01() < _mut_prob){
            X_new.isGW[i] = !X_new.isGW[i]; // Switch gene
        }
    }

	return X_new;
}

MySolution crossover(const MySolution& X1, const MySolution& X2,const function<double(void)> &rnd01) {
	MySolution X_new;
    init_genes(X_new, rnd01);
    unsigned int x_point = (unsigned int) (rnd01()*(double)_enddevices); // Crossover point
    for(unsigned int i = 0; i < x_point; i++){
        X_new.isGW[i] = X1.isGW[i];
    }
    for(unsigned int i = x_point; i < _enddevices; i++){
        X_new.isGW[i] = X2.isGW[i];
    }
	return X_new;
}
```

Now the output of ```SO_report_generation``` has more sense. I will leave this issue open until getting some feedback of this approach.

**Comment by matiasmicheletto** [comment]
Hi Arash, thanks for the quick response. I wasn't aware about struct destructors. This approach fits very well for my case.
Answering your question, I am using malloc instead of vector because I believe that for the case of binary data, simple arrays are faster and consume less memory than std::vector class.
I am happy with the current results, so I am closing the issue, and I hope some else find this useful. 
Again, thank you very much for your help!
